File: api/token.py
"""
Vercel Serverless Function for LiveKit Token Generation
This replaces the /api/token endpoint from server.py for Vercel deployment
"""

import os
import uuid
import time
import json
import jwt
from http import HTTPStatus


def handler(request):
    """Vercel serverless function handler for token generation."""
    
    # Only allow GET requests
    if request.method != 'GET':
        return {
            'statusCode': HTTPStatus.METHOD_NOT_ALLOWED,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'Method not allowed'})
        }
    
    # Load credentials from environment variables
    LIVEKIT_URL = os.getenv('LIVEKIT_URL')
    LIVEKIT_API_KEY = os.getenv('LIVEKIT_API_KEY')
    LIVEKIT_API_SECRET = os.getenv('LIVEKIT_API_SECRET')
    
    if not all([LIVEKIT_URL, LIVEKIT_API_KEY, LIVEKIT_API_SECRET]):
        return {
            'statusCode': HTTPStatus.INTERNAL_SERVER_ERROR,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'LiveKit credentials not configured'})
        }
    
    try:
        # Generate unique identity for this user
        identity = f"user-{uuid.uuid4().hex[:8]}"
        room_name = "cheeko-room"
        
        # Create JWT token manually using PyJWT
        # This matches the LiveKit token format
        now = int(time.time())
        payload = {
            "exp": now + 3600,  # Token expires in 1 hour
            "iss": LIVEKIT_API_KEY,
            "nbf": now - 10,  # Not before (with 10s buffer)
            "sub": identity,
            "name": "User",
            "video": {
                "roomJoin": True,
                "room": room_name,
                "canPublish": True,
                "canSubscribe": True,
            }
        }
        
        jwt_token = jwt.encode(payload, LIVEKIT_API_SECRET, algorithm="HS256")
        
        # Return successful response
        response_data = {
            'token': jwt_token,
            'url': LIVEKIT_URL,
            'identity': identity,
            'room': room_name,
        }
        
        return {
            'statusCode': HTTPStatus.OK,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps(response_data)
        }
        
    except Exception as e:
        return {
            'statusCode': HTTPStatus.INTERNAL_SERVER_ERROR,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': f'Token generation failed: {str(e)}'})
        }

File: api/token.js
const jwt = require('jsonwebtoken');

export default async function handler(req, res) {
    // CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    const LIVEKIT_URL = process.env.LIVEKIT_URL;
    const LIVEKIT_API_KEY = process.env.LIVEKIT_API_KEY;
    const LIVEKIT_API_SECRET = process.env.LIVEKIT_API_SECRET;

    if (!LIVEKIT_URL || !LIVEKIT_API_KEY || !LIVEKIT_API_SECRET) {
        return res.status(500).json({ error: 'LiveKit credentials not configured' });
    }

    try {
        const userDetails = req.body?.userDetails || {};
        const sessionId = Math.random().toString(36).substring(2, 10);
        const identity = `user-${sessionId}`;
        const roomName = `cheeko-room-${sessionId}`;  // Unique room per user!
        const userName = userDetails.name || 'User';

        const now = Math.floor(Date.now() / 1000);
        const payload = {
            exp: now + 3600,
            iss: LIVEKIT_API_KEY,
            nbf: now - 10,
            sub: identity,
            name: userName,
            video: {
                roomJoin: true,
                room: roomName,
                canPublish: true,
                canSubscribe: true,
            }
        };

        // Add metadata if provided
        if (userDetails && Object.keys(userDetails).length > 0) {
            payload.metadata = JSON.stringify(userDetails);
        }

        const token = jwt.sign(payload, LIVEKIT_API_SECRET, { algorithm: 'HS256' });

        return res.status(200).json({
            token,
            url: LIVEKIT_URL,
            identity,
            room: roomName,
        });
    } catch (error) {
        console.error('Token generation error:', error);
        return res.status(500).json({ error: 'Token generation failed' });
    }
}
